File: swig/python/test_mer_file.py
import jellyfish
import unittest
import sys
import os
from collections import Counter

class TestMerFile(unittest.TestCase):
    def setUp(self):
        self.mf = jellyfish.ReadMerFile(os.path.join(data, "swig_python.jf"))

    def test_histo(self):
        histo = Counter()
        while self.mf.next_mer():
            histo[self.mf.count()] += 1

        jf_histo = Counter()
        with open(os.path.join(data, "swig_python.histo")) as f:
            for line in f:
                num, count = [int(n) for n in line.split()]
                self.assertEqual(count, histo[num])

    def test_dump(self):
        good = True
        with open(os.path.join(data, "swig_python.dump")) as f:
            for line in f:
                good = good and self.mf.next_mer()
                if not good: break
                a = line.split()
                good = good and a[0] == str(self.mf.mer()) and int(a[1]) == self.mf.count()
                if not good: break
        self.assertTrue(good)

    def test_iter(self):
        good = True
        with open(os.path.join(data, "swig_python.dump")) as f:
            for mer, count in self.mf:
                line = f.readline()
                good = good and line
                if not good: break
                fmer, fcount = line.split()
                good = good and fmer == str(mer) and int(fcount) == count
                if not good: break
            self.assertTrue(good)
            line = f.readline()
            self.assertTrue(not line)

    def test_query(self):
        good = True
        qf   = jellyfish.QueryMerFile(os.path.join(data, "swig_python.jf"))
        for mer, count in self.mf:
            good = good and count == qf[mer]
            if not good: break
        self.assertTrue(good)

if __name__ == '__main__':
    data = sys.argv.pop(1)
    unittest.main()

File: swig/ruby/test_mer_file.rb
require 'minitest/autorun'
require 'jellyfish'

$data = ARGV.shift

class TestMerFile < MiniTest::Unit::TestCase
  def setup
    @mf = Jellyfish::ReadMerFile.new(File.join($data, "swig_ruby.jf"))
  end

  def test_histo
    histo = []
    histo[@mf.count] = (histo[@mf.count] || 0) + 1 while @mf.next_mer

    jf_histo = []
    open(File.join($data, "swig_ruby.histo")) { |f|
      f.each_line.each { |l|
        freq, count = l.split.map {|x| x.to_i }
        jf_histo[freq] = count
      }
    }
    
    assert_equal jf_histo, histo
  end

  def test_each
    open(File.join($data, "swig_ruby.dump")) { |f|
      @mf.each { |m, c|
        l = f.readline
        assert l
        fm, fc = l.split
        assert_equal fm, m.to_s
        assert_equal fc.to_i, c
      }
      assert_raises(EOFError) { f.readline }
    }
  end

  def test_dump
    open(File.join($data, "swig_ruby.dump")) { |f|
      f.each_line.each { |l|
        mer, count = l.split
        assert @mf.next_mer
        assert_equal(mer, @mf.mer.to_s)
        assert_equal(count.to_i, @mf.count)
      }
    }
    assert !@mf.next_mer
  end

  def test_query
    query = Jellyfish::QueryMerFile.new(File.join($data, "swig_ruby.jf"))
    @mf.each { |m, c|
      assert_equal c, query[m]
    }
  end
end
